src/project-converter.ts

```diff
@@ -0,0 +1,159 @@
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Document} from 'polymer-analyzer';
+
+import {ConversionSettings} from './conversion-settings';
+import {DocumentConverter} from './document-converter';
+import {ConversionResult, JsExport} from './js-module';
+import {ConvertedDocumentFilePath, OriginalDocumentUrl} from './urls/types';
+import {UrlHandlerInterface} from './urls/url-handler-interface';
+import {getDocumentUrl} from './urls/util';
+
+/**
+ * ProjectConverter provides the top-level interface for running a project
+ * conversion. convertDocument() should be called to kick off any conversion,
+ * and getResults() should be called once conversion is complete.
+ *
+ * For best results, only one ProjectConverter instance should be needed, so
+ * that it can cache results and avoid duplicate, extraneous document
+ * conversions.
+ *
+ * ProjectConverter is indifferent to the layout of the project, delegating any
+ * special URL handling/resolution to the urlHandler provided to the
+ * constructor.
+ */
+export class ProjectConverter {
+  readonly urlHandler: UrlHandlerInterface;
+  readonly conversionSettings: ConversionSettings;
+
+  /**
+   * A cache of all JS Exports by namespace, to map implicit HTML imports to
+   * explicit named JS imports.
+   */
+  readonly namespacedExports = new Map<string, JsExport>();
+  /**
+   * A cache of all converted documents. Document conversions should be
+   * idempotent, so conversion results can be safely cached.
+   */
+  readonly conversionResults = new Map<OriginalDocumentUrl, ConversionResult>();
+
+  constructor(
+      urlHandler: UrlHandlerInterface, conversionSettings: ConversionSettings) {
+    this.urlHandler = urlHandler;
+    this.conversionSettings = conversionSettings;
+  }
+
+  /**
+   * Convert a document and any of its dependencies. The output format (JS
+   * Module or HTML Document) is determined by whether the file is included in
+   * conversionSettings.includes.
+   */
+  convertDocument(document: Document) {
+    try {
+      this.conversionSettings.includes.has(document.url) ?
+          this.convertDocumentToJs(document, new Set()) :
+          this.convertDocumentToHtml(document, new Set());
+    } catch (e) {
+      console.error(`Error in ${document.url}`, e);
+    }
+  }
+
+  /**
+   * Check if a document is explicitly excluded or has already been converted
+   * to decide if it should be converted or skipped.
+   */
+  shouldConvertDocument(document: Document): boolean {
+    const documentUrl = getDocumentUrl(document);
+    if (this.conversionResults.has(documentUrl)) {
+      return false;
+    }
+    for (const exclude of this.conversionSettings.excludes) {
+      if (documentUrl.endsWith(exclude)) {
```

Comment:
Is the format of the `exclude` string such that it is always preceded by a delimiter like '/' such that `endsWith` will reliably match `apples` to `my/apples` but not `crapples`?

Reply:
This is similar to https://github.com/Polymer/polymer-modulizer/pull/224#discussion_r150687180. `endsWith()` is a more flexible matcher (it works on both package and workspace layouts) but I agree that it might be too flexible. Lets make this an exact `===` matcher for now, and I'll add an issue to convert to something thats both explicit & flexible, like glob matching.